src/expr/impl/src/aggregate/jsonb_agg.rs

```diff
@@ -36,15 +35,130 @@ fn jsonb_agg(state: Option<JsonbVal>, input: Option<impl Into<Value>>) -> JsonbV
 #[aggregate("jsonb_object_agg(varchar, varchar) -> jsonb")]
 #[aggregate("jsonb_object_agg(varchar, jsonb) -> jsonb")]
 fn jsonb_object_agg(
-    state: Option<JsonbVal>,
+    state: &mut JsonbObjectState,
     key: Option<&str>,
-    value: Option<impl Into<Value>>,
-) -> Result<JsonbVal> {
+    value: Option<impl ToJson>,
+) -> Result<()> {
     let key = key.ok_or(ExprError::FieldNameNull)?;
-    let mut jsonb = state.unwrap_or_else(|| Value::Object(Default::default()).into());
-    match jsonb.as_serde_mut() {
-        Value::Object(map) => map.insert(key.into(), value.map_or(Value::Null, Into::into)),
-        _ => unreachable!("invalid jsonb state"),
-    };
-    Ok(jsonb)
+    state.0.add_string(key);
+    match value {
+        Some(value) => value.add_to(&mut state.0),
+        None => state.0.add_null(),
+    }
+    Ok(())
+}
+
+#[derive(Debug)]
+struct JsonbArrayState(jsonbb::Builder);
+
+impl EstimateSize for JsonbArrayState {
+    fn estimated_heap_size(&self) -> usize {
+        self.0.capacity()
+    }
+}
+
+impl AggStateDyn for JsonbArrayState {}
+
+/// Creates an initial state.
+impl Default for JsonbArrayState {
+    fn default() -> Self {
+        let mut builder = jsonbb::Builder::default();
+        builder.begin_array();
+        Self(builder)
+    }
+}
+
+/// Finishes aggregation and returns the result.
+impl From<&JsonbArrayState> for ScalarImpl {
+    fn from(builder: &JsonbArrayState) -> Self {
+        // TODO: avoid clone
+        let mut builder = builder.0.clone();
+        builder.end_array();
+        let jsonb: JsonbVal = builder.finish().into();
+        jsonb.into()
+    }
+}
+
+#[derive(Debug)]
+struct JsonbObjectState(jsonbb::Builder);
+
+impl EstimateSize for JsonbObjectState {
+    fn estimated_heap_size(&self) -> usize {
+        self.0.capacity()
+    }
+}
+
+impl AggStateDyn for JsonbObjectState {}
+
+/// Creates an initial state.
+impl Default for JsonbObjectState {
+    fn default() -> Self {
+        let mut builder = jsonbb::Builder::default();
+        builder.begin_object();
+        Self(builder)
+    }
+}
+
+/// Finishes aggregation and returns the result.
+impl From<&JsonbObjectState> for ScalarImpl {
+    fn from(builder: &JsonbObjectState) -> Self {
+        // TODO: avoid clone
+        let mut builder = builder.0.clone();
+        builder.end_object();
+        let jsonb: JsonbVal = builder.finish().into();
+        jsonb.into()
+    }
+}
+
+/// Values that can be converted to JSON.
+trait ToJson {
+    fn add_to(self, builder: &mut jsonbb::Builder);
+}
```

Comment:
The `&mut jsonbb::Builder` parameter reminds me of this. Eventually we shall also accept BytesWriter and jsonbb Builder as expr arguments so that `#[function]` returning `bytea` and `jsonb` can write their output to destination directly.

https://github.com/risingwavelabs/risingwave/blob/39f71a171b8a4cece71902dde973ff58d4215fcf/src/common/src/array/utf8_array.rs#L175-L177

cc @KeXiangWang for #12834